// @connext/smart-contracts/dist/src/typechain-types/factories/contracts/messaging/interfaces/ambs/Multichain__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Multichain__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "_to",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "_data",
                type: "bytes",
            },
            {
                internalType: "address",
                name: "_fallback",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "_toChainID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_flags",
                type: "uint256",
            },
        ],
        name: "anyCall",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "_appID",
                type: "string",
            },
            {
                internalType: "uint256",
                name: "_toChainID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "_dataLength",
                type: "uint256",
            },
        ],
        name: "calcSrcFees",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "context",
        outputs: [
            {
                internalType: "address",
                name: "from",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "fromChainID",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "nonce",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "executor",
        outputs: [
            {
                internalType: "address",
                name: "executor",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class Multichain__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.Multichain__factory = Multichain__factory;
Multichain__factory.abi = _abi;
