// @connext/nxtp-utils/dist/logging/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const pino_1 = __importDefault(require("pino"));
const _1 = require(".");
/**
 * @classdesc Designed to log information in a uniform way to make parsing easier
 */
class Logger {
    constructor(opts, forcedLevel) {
        this.opts = opts;
        this.forcedLevel = forcedLevel;
        this.log = (0, pino_1.default)(this.opts);
    }
    child(bindings, forcedLevel) {
        return new Logger({ ...this.opts, ...bindings }, forcedLevel);
    }
    debug(msg, requestContext, methodContext, ctx) {
        this.print(this.forcedLevel ?? "debug", requestContext, methodContext, this.forcedLevel ? { ...ctx, intendedLevel: "debug" } : ctx, msg);
    }
    info(msg, requestContext, methodContext, ctx) {
        this.print(this.forcedLevel ?? "info", requestContext, methodContext, this.forcedLevel ? { ...ctx, intendedLevel: "info" } : ctx, msg);
    }
    warn(msg, requestContext, methodContext, ctx) {
        this.print(this.forcedLevel ?? "warn", requestContext, methodContext, this.forcedLevel ? { ...ctx, intendedLevel: "warn" } : ctx, msg);
    }
    error(msg, requestContext, methodContext, error, ctx) {
        this.print(this.forcedLevel ?? "error", requestContext, methodContext, this.forcedLevel ? { ...ctx, error, intendedLevel: "error" } : { ...ctx, error }, msg);
    }
    print(level, requestContext = (0, _1.createRequestContext)("Logger.print"), methodContext = (0, _1.createMethodContext)("Logger.print"), ctx = {}, msg) {
        return this.log[level]({ requestContext, methodContext, ...ctx }, msg);
    }
}
exports.Logger = Logger;
