// @connext/smart-contracts/dist/src/typechain-types/factories/contracts/messaging/libraries/MerkleLib__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MerkleLib__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "MerkleLib__insert_treeIsFull",
        type: "error",
    },
];
const _bytecode = "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212208f03b44d10acfd54ee50c9550572558998a89d47e47c8182c01f7c8b62bdb42364736f6c63430008110033";
const isSuperArgs = (xs) => xs.length > 1;
class MerkleLib__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.MerkleLib__factory = MerkleLib__factory;
MerkleLib__factory.bytecode = _bytecode;
MerkleLib__factory.abi = _abi;
