// @connext/smart-contracts/dist/src/typechain-types/factories/@matterlabs/zksync-contracts/l1/contracts/zksync/libraries/Diamond__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Diamond__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "facet",
                        type: "address",
                    },
                    {
                        internalType: "enum Diamond.Action",
                        name: "action",
                        type: "uint8",
                    },
                    {
                        internalType: "bool",
                        name: "isFreezable",
                        type: "bool",
                    },
                    {
                        internalType: "bytes4[]",
                        name: "selectors",
                        type: "bytes4[]",
                    },
                ],
                indexed: false,
                internalType: "struct Diamond.FacetCut[]",
                name: "facetCuts",
                type: "tuple[]",
            },
            {
                indexed: false,
                internalType: "address",
                name: "initAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "initCalldata",
                type: "bytes",
            },
        ],
        name: "DiamondCut",
        type: "event",
    },
];
const _bytecode = "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220fe62b76397b91d8a8051e4652d559ff6e10ab834c7cbda53fcebaeea60c7e32064736f6c63430008110033";
const isSuperArgs = (xs) => xs.length > 1;
class Diamond__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.Diamond__factory = Diamond__factory;
Diamond__factory.bytecode = _bytecode;
Diamond__factory.abi = _abi;
