// @connext/smart-contracts/dist/src/typechain-types/factories/contracts/core/connext/helpers/RelayerProxy.sol/ISpokeConnector__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ISpokeConnector__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "bytes",
                        name: "message",
                        type: "bytes",
                    },
                    {
                        internalType: "bytes32[32]",
                        name: "path",
                        type: "bytes32[32]",
                    },
                    {
                        internalType: "uint256",
                        name: "index",
                        type: "uint256",
                    },
                ],
                internalType: "struct ISpokeConnector.Proof[]",
                name: "_proofs",
                type: "tuple[]",
            },
            {
                internalType: "bytes32",
                name: "_aggregateRoot",
                type: "bytes32",
            },
            {
                internalType: "bytes32[32]",
                name: "_aggregatePath",
                type: "bytes32[32]",
            },
            {
                internalType: "uint256",
                name: "_aggregateIndex",
                type: "uint256",
            },
        ],
        name: "proveAndProcess",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "_encodedData",
                type: "bytes",
            },
        ],
        name: "send",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
];
class ISpokeConnector__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ISpokeConnector__factory = ISpokeConnector__factory;
ISpokeConnector__factory.abi = _abi;
