// @connext/smart-contracts/dist/src/typechain-types/factories/contracts/core/connext/libraries/AssetLogic__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetLogic__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "AssetLogic__getConfig_notRegistered",
        type: "error",
    },
    {
        inputs: [],
        name: "AssetLogic__getTokenIndexFromStableSwapPool_notExist",
        type: "error",
    },
    {
        inputs: [],
        name: "AssetLogic__handleIncomingAsset_feeOnTransferNotSupported",
        type: "error",
    },
    {
        inputs: [],
        name: "AssetLogic__handleIncomingAsset_nativeAssetNotSupported",
        type: "error",
    },
    {
        inputs: [],
        name: "AssetLogic__handleOutgoingAsset_notNative",
        type: "error",
    },
    {
        inputs: [],
        name: "AssetLogic__swapAsset_externalStableSwapPoolDoesNotExist",
        type: "error",
    },
];
const _bytecode = "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212209e978a0cdf3084c52c4fd1673e22aa7a086ea102387ff29a97d49c1dc0e3865164736f6c63430008110033";
const isSuperArgs = (xs) => xs.length > 1;
class AssetLogic__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.AssetLogic__factory = AssetLogic__factory;
AssetLogic__factory.bytecode = _bytecode;
AssetLogic__factory.abi = _abi;
