// @connext/smart-contracts/dist/src/typechain-types/factories/contracts/core/connext/interfaces/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IXReceiver__factory = exports.IStableSwap__factory = exports.IPriceOracle__factory = exports.IERC165__factory = exports.IDiamondLoupe__factory = exports.IDiamondCut__factory = exports.IConnext__factory = exports.IBridgeToken__factory = exports.IAavePool__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
var IAavePool__factory_1 = require("./IAavePool__factory");
Object.defineProperty(exports, "IAavePool__factory", { enumerable: true, get: function () { return IAavePool__factory_1.IAavePool__factory; } });
var IBridgeToken__factory_1 = require("./IBridgeToken__factory");
Object.defineProperty(exports, "IBridgeToken__factory", { enumerable: true, get: function () { return IBridgeToken__factory_1.IBridgeToken__factory; } });
var IConnext__factory_1 = require("./IConnext__factory");
Object.defineProperty(exports, "IConnext__factory", { enumerable: true, get: function () { return IConnext__factory_1.IConnext__factory; } });
var IDiamondCut__factory_1 = require("./IDiamondCut__factory");
Object.defineProperty(exports, "IDiamondCut__factory", { enumerable: true, get: function () { return IDiamondCut__factory_1.IDiamondCut__factory; } });
var IDiamondLoupe__factory_1 = require("./IDiamondLoupe__factory");
Object.defineProperty(exports, "IDiamondLoupe__factory", { enumerable: true, get: function () { return IDiamondLoupe__factory_1.IDiamondLoupe__factory; } });
var IERC165__factory_1 = require("./IERC165__factory");
Object.defineProperty(exports, "IERC165__factory", { enumerable: true, get: function () { return IERC165__factory_1.IERC165__factory; } });
var IPriceOracle__factory_1 = require("./IPriceOracle__factory");
Object.defineProperty(exports, "IPriceOracle__factory", { enumerable: true, get: function () { return IPriceOracle__factory_1.IPriceOracle__factory; } });
var IStableSwap__factory_1 = require("./IStableSwap__factory");
Object.defineProperty(exports, "IStableSwap__factory", { enumerable: true, get: function () { return IStableSwap__factory_1.IStableSwap__factory; } });
var IXReceiver__factory_1 = require("./IXReceiver__factory");
Object.defineProperty(exports, "IXReceiver__factory", { enumerable: true, get: function () { return IXReceiver__factory_1.IXReceiver__factory; } });
