// @connext/smart-contracts/dist/src/typechain-types/factories/contracts/messaging/interfaces/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IRootManager__factory = exports.IOutbox__factory = exports.IMessageRecipient__factory = exports.IHubConnector__factory = exports.IConnectorManager__factory = exports.IConnector__factory = exports.ambs = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
exports.ambs = __importStar(require("./ambs"));
var IConnector__factory_1 = require("./IConnector__factory");
Object.defineProperty(exports, "IConnector__factory", { enumerable: true, get: function () { return IConnector__factory_1.IConnector__factory; } });
var IConnectorManager__factory_1 = require("./IConnectorManager__factory");
Object.defineProperty(exports, "IConnectorManager__factory", { enumerable: true, get: function () { return IConnectorManager__factory_1.IConnectorManager__factory; } });
var IHubConnector__factory_1 = require("./IHubConnector__factory");
Object.defineProperty(exports, "IHubConnector__factory", { enumerable: true, get: function () { return IHubConnector__factory_1.IHubConnector__factory; } });
var IMessageRecipient__factory_1 = require("./IMessageRecipient__factory");
Object.defineProperty(exports, "IMessageRecipient__factory", { enumerable: true, get: function () { return IMessageRecipient__factory_1.IMessageRecipient__factory; } });
var IOutbox__factory_1 = require("./IOutbox__factory");
Object.defineProperty(exports, "IOutbox__factory", { enumerable: true, get: function () { return IOutbox__factory_1.IOutbox__factory; } });
var IRootManager__factory_1 = require("./IRootManager__factory");
Object.defineProperty(exports, "IRootManager__factory", { enumerable: true, get: function () { return IRootManager__factory_1.IRootManager__factory; } });
