// @connext/smart-contracts/dist/src/typechain-types/factories/contracts/messaging/interfaces/IRootManager__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IRootManager__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint32",
                name: "_domain",
                type: "uint32",
            },
            {
                internalType: "bytes32",
                name: "_outbound",
                type: "bytes32",
            },
        ],
        name: "aggregate",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address[]",
                name: "_connectors",
                type: "address[]",
            },
            {
                internalType: "uint256[]",
                name: "_fees",
                type: "uint256[]",
            },
            {
                internalType: "bytes[]",
                name: "_encodedData",
                type: "bytes[]",
            },
        ],
        name: "propagate",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
];
class IRootManager__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IRootManager__factory = IRootManager__factory;
IRootManager__factory.abi = _abi;
