// @connext/smart-contracts/dist/src/typechain-types/factories/contracts/messaging/interfaces/ambs/arbitrum/ArbitrumL2Amb__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArbitrumL2Amb__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "destAddr",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "calldataForL1",
                type: "bytes",
            },
        ],
        name: "sendTxToL1",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
];
class ArbitrumL2Amb__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ArbitrumL2Amb__factory = ArbitrumL2Amb__factory;
ArbitrumL2Amb__factory.abi = _abi;
