// @connext/smart-contracts/dist/src/typechain-types/factories/@matterlabs/zksync-contracts/l1/contracts/zksync/interfaces/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IZkSync__factory = exports.IMailbox__factory = exports.IGovernance__factory = exports.IGetters__factory = exports.IExecutor__factory = exports.IDiamondCut__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
var IDiamondCut__factory_1 = require("./IDiamondCut__factory");
Object.defineProperty(exports, "IDiamondCut__factory", { enumerable: true, get: function () { return IDiamondCut__factory_1.IDiamondCut__factory; } });
var IExecutor__factory_1 = require("./IExecutor__factory");
Object.defineProperty(exports, "IExecutor__factory", { enumerable: true, get: function () { return IExecutor__factory_1.IExecutor__factory; } });
var IGetters__factory_1 = require("./IGetters__factory");
Object.defineProperty(exports, "IGetters__factory", { enumerable: true, get: function () { return IGetters__factory_1.IGetters__factory; } });
var IGovernance__factory_1 = require("./IGovernance__factory");
Object.defineProperty(exports, "IGovernance__factory", { enumerable: true, get: function () { return IGovernance__factory_1.IGovernance__factory; } });
var IMailbox__factory_1 = require("./IMailbox__factory");
Object.defineProperty(exports, "IMailbox__factory", { enumerable: true, get: function () { return IMailbox__factory_1.IMailbox__factory; } });
var IZkSync__factory_1 = require("./IZkSync__factory");
Object.defineProperty(exports, "IZkSync__factory", { enumerable: true, get: function () { return IZkSync__factory_1.IZkSync__factory; } });
