// @connext/smart-contracts/dist/src/typechain-types/factories/@matterlabs/zksync-contracts/l2/system-contracts/interfaces/IAccountCodeStorage__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IAccountCodeStorage__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint256",
                name: "_input",
                type: "uint256",
            },
        ],
        name: "getCodeHash",
        outputs: [
            {
                internalType: "bytes32",
                name: "codeHash",
                type: "bytes32",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "_input",
                type: "uint256",
            },
        ],
        name: "getCodeSize",
        outputs: [
            {
                internalType: "uint256",
                name: "codeSize",
                type: "uint256",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_address",
                type: "address",
            },
        ],
        name: "getRawCodeHash",
        outputs: [
            {
                internalType: "bytes32",
                name: "codeHash",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_address",
                type: "address",
            },
            {
                internalType: "bytes32",
                name: "_hash",
                type: "bytes32",
            },
        ],
        name: "storeCodeHash",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IAccountCodeStorage__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IAccountCodeStorage__factory = IAccountCodeStorage__factory;
IAccountCodeStorage__factory.abi = _abi;
