// @connext/smart-contracts/dist/src/typechain-types/factories/@matterlabs/zksync-contracts/l2/system-contracts/interfaces/IImmutableSimulator__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IImmutableSimulator__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "_dest",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "_index",
                type: "uint256",
            },
        ],
        name: "getImmutable",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_dest",
                type: "address",
            },
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "index",
                        type: "uint256",
                    },
                    {
                        internalType: "bytes32",
                        name: "value",
                        type: "bytes32",
                    },
                ],
                internalType: "struct ImmutableData[]",
                name: "immutables",
                type: "tuple[]",
            },
        ],
        name: "setImmutables",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IImmutableSimulator__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IImmutableSimulator__factory = IImmutableSimulator__factory;
IImmutableSimulator__factory.abi = _abi;
