// @connext/smart-contracts/dist/src/typechain-types/factories/@matterlabs/zksync-contracts/l2/system-contracts/interfaces/IKnownCodesStorage__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IKnownCodesStorage__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "bytecodeHash",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "bool",
                name: "sendBytecodeToL1",
                type: "bool",
            },
        ],
        name: "MarkedAsKnown",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "_hash",
                type: "bytes32",
            },
        ],
        name: "getMarker",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bool",
                name: "_shouldSendToL1",
                type: "bool",
            },
            {
                internalType: "bytes32[]",
                name: "_hashes",
                type: "bytes32[]",
            },
        ],
        name: "markFactoryDeps",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IKnownCodesStorage__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IKnownCodesStorage__factory = IKnownCodesStorage__factory;
IKnownCodesStorage__factory.abi = _abi;
