// @connext/smart-contracts/dist/src/typechain-types/factories/@matterlabs/zksync-contracts/l2/system-contracts/interfaces/IL1Messenger__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IL1Messenger__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "_sender",
                type: "address",
            },
            {
                indexed: true,
                internalType: "bytes32",
                name: "_hash",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "_message",
                type: "bytes",
            },
        ],
        name: "L1MessageSent",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "_message",
                type: "bytes",
            },
        ],
        name: "sendToL1",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IL1Messenger__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IL1Messenger__factory = IL1Messenger__factory;
IL1Messenger__factory.abi = _abi;
